src/lib/core/common-behaviors/id.ts

```
@@ -0,0 +1,35 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Constructor} from './constructor';
+
+/** Monotonically increasing integer used to auto-generate unique ids. */
+let nextUniqueId = 0;
+
+/** @docs-private */
+export interface CanId {
+  id: string;
+}
+
+/**
+ * Mixin to augment a directive with an `id` property.
+ * The id property will fall back to auto-generated unique ids if developers don't specify one.
+ */
+export function mixinId<T extends Constructor<{}>>(base: T, componentName: string)
```

I'm not sure this is a good idea. I'd consider returning a different ID than what the user defines to be surprising / confusing. For components that need an ID, I'd rather throw an error if its set to null.